src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3757,45 +3760,93 @@ private void LearnFromEqualsMethod(MethodSymbol method, BoundCall node, TypeWith
                 {
-                    return true;
+                    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
+                    if (implementationMethod is null)
+                    {
+                        // we know no base type will implement this interface member either
+                        return false;
+                    }
+
+                    // 'method' definitely won't match if the implementation is explict.
+                    if (!implementationMethod.IsExplicitInterfaceImplementation())
+                    {
+                        // could be calling an override of a method that implements the interface method
+                        for (var overriddenMethod = method; overriddenMethod is object; overriddenMethod = overriddenMethod.OverriddenMethod)
+                        {
+                            if (overriddenMethod.Equals(implementationMethod))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+
+                    // the Equals method being called isn't the method that implements the interface method in this type.
+                    // it could be a method that implements the interface on a base type, so check again with the base type of 'implementationMethod.ContainingType'
+
+                    // e.g. in this hierarchy:
+                    // class A -> B -> C -> D
+                    // method virtual B.Equals -> override D.Equals
+                    //
+                    // we would potentially check:
+                    // 1. D.Equals when called on D, then B.Equals when called on D
+                    // 2. B.Equals when called on C
+                    // 3. B.Equals when called on B
+                    // 4. give up when checking A, since B.Equals is not overriding anything in A
+
+                    // we know that implementationMethod.ContainingType is the same type or a base type of 'baseType',
+                    // and that the implementation method will be the same between 'baseType' and 'implementationMethod.ContainingType'.
+                    // we step through the intermediate bases in order to skip unnecessary override methods.
+                    while (!baseType.Equals(implementationMethod.ContainingType))
+                    {
+                        baseType = baseType.BaseTypeNoUseSiteDiagnostics;
+                        // the implementation method must be contained in this 'baseType' or one of its bases.
+                        Debug.Assert(baseType is object);
+
+                        // since we're about to move on to the base of 'method.ContainingType',
+                        // we know the implementation could only be an overridden method of 'method'.
+                        if (baseType.Equals(method.ContainingType))
```

Comment:
>method [](start = 44, length = 6)

It looks like this can cause a NullReferenceException, if ```method``` becomes null on the previous iteration. Please add a test for a scenario like that.

Reply:
Done